sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -797,6 +831,14 @@ object NullPropagation extends Rule[LogicalPlan] {
       // a null literal.
       case e: NullIntolerant if e.children.exists(isNullLiteral) =>
         Literal.create(null, e.dataType)
+
+      // Using IsNull(e(inputs)) == IsNull(input1) or IsNull(input2) ... rules
+      // E.g. IsNull(Not(null)) == IsNull(null)
+      // E.g. IsNotNull(a === b) == And(IsNotNull(a), IsNotNull(b))
+      case IsNull(e: NullIntolerant) if !e.isInstanceOf[ExtractValue] && e.children.nonEmpty =>
```

Comment:
Why is `ExtractValue` excluded?

Reply:
`ExtractValue` is used for the query planner to extract the field. E.g. the planner resolves column `a` to `a#123`
We cannot optimize `IsNull(a#123)` case.
I will add this explanation to the comment.

Reply:
updated the comment